### PR TITLE
Compute manipulator Jacobians for complex mimic joints by chain rule

### PR DESCRIPTION
**Pull request by tgn3000**
- Aimed to handle Jacobian calculations where a mimic joint depends on other mimic joint(s). Dependencies of active joints and mimic joints form a directed acyclic graph.
- Modify, polish Jacobian computation functions, such as `ComputeJacobianAxisAngle`, `ComputeJacobianTranslation`, `CalculateRotationJacobian`, so that they correctly propagate the partial derivatives by chain rule.
- Document the methods in code that set up the joint dependencies, parse the function evaluations in a xml, and propagate the numerical results.
- Add log code for debugging.

### CODE REVIEW

File: include/openrave/kinbody.h [hunk 1]
```diff
@@ -854,15 +854,20 @@ class OPENRAVE_API KinBody : public InterfaceBase
             bool operator ==(const DOFFormat& r) const;
             boost::shared_ptr<Joint> GetJoint(KinBody &parent) const;
             boost::shared_ptr<Joint const> GetJoint(const KinBody &parent) const;
+            std::string to_string() const;
```
Comment (rdiankov) [inline include/openrave/kinbody.h:857]: @tgn3000 what's the difference between to_string and getting the JSON struct and dumping it? You want to display shorter information? If so, what's the deciding factor for this shorter info?
Can we also give the function a nicer name following mujin conventions?
thanks

File: src/libopenrave/kinbody.cpp [hunk 2]
```diff
@@ -3467,84 +3539,117 @@ void KinBody::_ComputeInternalInformation()
...
     try {
         // initialize all the mimic equations
-        for(int ijoints = 0; ijoints < 2; ++ijoints) {
-            vector<JointPtr>& vjoints = ijoints ? _vPassiveJoints : _vecjoints;
-            FOREACH(itjoint,vjoints) {
-                for(int i = 0; i < (*itjoint)->GetDOF(); ++i) {
-                    if( !!(*itjoint)->_vmimic[i] ) {
-                        std::string poseq = (*itjoint)->_vmimic[i]->_equations[0], veleq = (*itjoint)->_vmimic[i]->_equations[1], acceleq = (*itjoint)->_vmimic[i]->_equations[2]; // have to copy since memory can become invalidated
-                        (*itjoint)->SetMimicEquations(i,poseq,veleq,acceleq);
+        for(bool bPassiveJoints : {false, true}) {
```
Comment (rdiankov) [inline src/libopenrave/kinbody.cpp:3542]: @tgn3000 i'm not sure all openrave users are using c++ with this fancy feature enabled...

File: src/libopenrave/kinbody.cpp [hunk 3]
```diff
@@ -3467,84 +3539,117 @@ void KinBody::_ComputeInternalInformation()
...
         while(bchanged) {
             bchanged = false;
-            FOREACH(itmimic,mapmimic) {
-                boost::shared_ptr<Mimic> mimic = itmimic->second;
-                Mimic::DOFHierarchy h;
-                h.dofformatindex = 0;
-                FOREACH(itdofformat,mimic->_vdofformat) {
-                    if( mapmimic.find(*itdofformat) == mapmimic.end() ) {
-                        continue; // this is normal, just means that the parent is a regular dof
+            for(const std::pair<const Mimic::DOFFormat, MimicPtr>& keyvalue : mapmimic) {
```
Comment (rdiankov) [inline src/libopenrave/kinbody.cpp:3604]: @tgn3000 i'm not sure openrave users are using a compiler with this fancy feature enabled...

File: src/libopenrave/kinbodyjoint.cpp [hunk 4]
```diff
@@ -1817,83 +1845,156 @@ void KinBody::Joint::SetMimicEquations(int iaxis, const std::string& poseq, cons
...
+    const MimicPtr& pmimic = _vmimic[iaxis];
+    const std::vector<Mimic::DOFFormat>& vdofformats = pmimic->_vdofformat; ///< collection of information of joints
+
+    std::vector<dReal> vDependedJointValues; ///< collect values of joints on which this joint *directly* depends on
+    for(const Mimic::DOFFormat& dofformat : vdofformats) {
+        const JointConstPtr dependedjoint = dofformat.GetJoint(*parent); ///< say joint y
+        vDependedJointValues.push_back(dependedjoint->GetValue(dofformat.axis));
+    }
+
+#if defined(DEBUG_KINBODYJOINT_CALCULATEJACOBIAN)
```
Comment (rdiankov) [inline src/libopenrave/kinbodyjoint.cpp:1941]: @tgn3000 can't you just do `if( IS_DEBUGLEVEL(Level_Verbose) )`?

File: src/libopenrave/kinbody.cpp (hunk 2 above)
Comment (tgn3000) [reply src/libopenrave/kinbody.cpp:3542]: This is just c++11 for-loop. In `CMakeLists.txt` we already had
```
set(CMAKE_CXX_STANDARD 11)
```

File: src/libopenrave/kinbody.cpp (hunk 3 above)
Comment (tgn3000) [reply src/libopenrave/kinbody.cpp:3604]: This is just c++11 for-loop. In `CMakeLists.txt` we already had
```
set(CMAKE_CXX_STANDARD 11)
```

File: src/libopenrave/kinbodyjoint.cpp (hunk 4 above)
Comment (tgn3000) [reply src/libopenrave/kinbodyjoint.cpp:1941]: ok

File: include/openrave/kinbody.h (hunk 1 above)
Comment (tgn3000) [reply include/openrave/kinbody.h]: Reduced log and removed `to_string()` related stuff. Tried JSON thing; did not work.

This is `kinbody.h`, so cannot `#include <openrave/openravejson.h>` to inherit `JsonSerializable`. Without inheriting, I cannot use `SaveJsonValue` unless applying the SFINAE technique. Tried and it worked, but slightly fancy.

**Comment by tgn3000** [comment]
clean and tested on TC:
https://teamcity.mujin.co.jp/viewLog.html?buildId=1681973

**Comment by sirop** [comment]
>     * Aimed to handle Jacobian calculations where a mimic joint depends on other mimic joint(s). Dependencies of active joints and mimic joints form a directed acyclic graph.
> 
>     * Modify, polish Jacobian computation functions, such as `ComputeJacobianAxisAngle`, `ComputeJacobianTranslation`, `CalculateRotationJacobian`, so that they correctly propagate the partial derivatives by chain rule.
> 
>     * Document the methods in code that set up the joint dependencies, parse the function evaluations in a xml, and propagate the numerical results.
> 
>     * Add log code for debugging.

Hallo.
Does this patch solve https://github.com/rdiankov/openrave/issues/154
or does it at least add Jacobian computation functions for parallel-link manipulators?

For instance, is such a case https://www.iri.upc.edu/research/webprojects/cuikweb/CuikSuite3-Doc/html/dir_3a7dc4066e4fb974c13ef5f31999df40.html now possible with OpenRAVE?

Thanks in advance for your answer.

**Comment by tgn3000** [comment]
@sirop Hello! Unfortunately these changes are irrelevant to inverse kinematics solvers or ikfast.
They aim to compute manipulator Jacobians correctly for manipulators that involve complex mimic joints with complex dependencies (which form a DAG). However, since we don't have many such examples for investigation or test, so could not guarantee much. 

I welcome if you could provide such robot examples/models for us to check, or tell us whether they compute correct results for you on your models.

Thank you for your comments!

File: src/libopenrave/kinbodyjoint.cpp [hunk 5]
```diff
@@ -1921,128 +1920,160 @@ void KinBody::Joint::GetMimicDOFIndices(std::vector<int>& vmimicdofs, int iaxis)
...
     }
-    KinBodyPtr parent(_parent);
-    std::vector<std::string> resultVars;
-    MimicPtr mimic(new Mimic());
-    mimic->_equations.at(0) = poseq;
-    mimic->_equations.at(1) = veleq;
-    mimic->_equations.at(2) = acceleq;
+
+    const MimicPtr pmimic(new Mimic());
+    pmimic->_equations = {poseq, veleq, acceleq}; ///< joint value, partial derivatives w.r.t depended joints, second-order derivatives (?)
```
Comment (tgn3000) [inline src/libopenrave/kinbodyjoint.cpp:1928]: revert this

**Comment by rdiankov** [comment]
@tgn3000 let me know when your new cleanup is ready

**Comment by rdiankov** [comment]
it took some time to waddle through the changes ;0)
thanks~

**Comment by tgn3000** [comment]
I was having fun. Thank you!

**Pull request merged**